File: 5_Longest_Palindromic_Substring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) return "";
        int start = 0, end = 0;
        for (int i=0; i<s.length(); i++){
            int len1 = expandMiddle(s, i, i); //for racecar
            int len2 = expandMiddle(s, i, i+1); // for normal cases
            int len = Math.max(len1, len2);
            if(len > end - start){
                start = i - ((len - 1) / 2);
                end = i + ((len) / 2);
            }
        }
        return s.substring(start, end + 1);
        
    }
    public int expandMiddle(String s, int left, int right){
        if (s == null || left > right) 
            return 0;
        while(left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)){
            left --;
            right ++;
        }
        return right - left - 1;
    }
}

File: 5_Longest_Palindromic_Substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int best_length = 0;
        string best_s = "";
        int n = s.length();
        for(int mid = 0; mid < n; mid++){
            for(int x = 0; mid - x >= 0 && mid + x < n; x++){
                if(s[mid - x] != s[mid + x])
                    break;
                int length = 2 * x + 1;
                if(length > best_length){
                    best_length = length;
                    best_s = s.substr(mid - x, length);
                }
            }
        }
        
        for(int mid = 0; mid < n - 1; mid++){
            for(int x = 1; mid - x + 1 >= 0 && mid + x < n; x++){
                if(s[mid - x + 1] != s[mid + x])
                    break;
                int length = 2 * x;
                if(length > best_length){
                    best_length = length;
                    best_s = s.substr(mid - x + 1, length);
                }
            }
        }
        return best_s;
    }
};
